pkg/controllers/nodeclass/controller.go

```diff
@@ -80,6 +87,13 @@ func (c *Controller) Reconcile(ctx context.Context, nodeClass *v1beta1.EC2NodeCl
 	return reconcile.Result{RequeueAfter: 5 * time.Minute}, err
 }
 
+func (c *Controller) Finalize(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) (reconcile.Result, error) {
+	if !controllerutil.ContainsFinalizer(nodeClass, v1beta1.TerminationFinalizer) {
+		return reconcile.Result{}, nil
+	}
+	return reconcile.Result{}, c.instanceProfileProvider.Delete(ctx, nodeClass)
```

Comment:
Does this fail if there are other instances using the same instance profile but owned by a different node class? That would require user intervention if I understand correctly, and is still possible even if we suffix each instance profile we generate with the name of the node class.

Reply:
Thinking further, since IAM is global, any clusters that have the same nodeclass names/setups will also be relying on this instance profile generation. Can we be sure the instance profile is safe to delete here if other instances are using it? Same with creation. Is this idempotent?

Reply:
> since IAM is global

That's a good callout. This should be region prefixed so that we are not hitting conflicts with other regions when creating deleting this instance profile for this role. In terms of the safety of deleting this instance profile if there are instances using it, I think it probably makes sense to hold EC2NodeClass deletion until all instances are dissociated from the instance profile.

I think there should be an implicit assumption here that the instance profile that is being generated here is for the explicit use and management of the Karpenter controller. You _can_ attach other instance types to the instance profile, but (if you do so), you will get failures when we try to remove the EC2NodeClass, which you will see inside of your error logs on the termination and you won't be able to fully remove the EC2NodeClass until you resolve those instance association errors (which you would have to deal with by doing an instance deletion)

Reply:
> This should be region prefixed so that we are not hitting conflicts with other regions when creating deleting this instance profile for this role. 

Can we make sure that the limit on instanceProfiles here is sufficiently large? IIUC from this change, Karpenter will now create/update/delete instanceProfiles for the number of EC2NodeClasses for an account across all regions. I can imagine for users with multiple clusters per region and multiple EC2NodeClasses per cluster that this may be a lot of instance profiles.

Reply:
> you will get failures when we try to remove the EC2NodeClass, which you will see inside of your error logs on the termination and you won't be able to fully remove the EC2NodeClass until you resolve those instance association errors (which you would have to deal with by doing an instance deletion)

Is there someway we can limit use of this instanceProfile? It seems odd that if a user created an instance (not necessarily connected to a cluster) with the instanceProfile that it would prevent cascade deletion of the nodeClass. Also having this IAM dependency now means that we can't do any type of Forceful deletion of this object without orphaning/leaking resources, which may be a sharp edge for karpenter uninstallations or cluster deletes.